plugins/outputs/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,293 @@
+func (s *Stackdriver) Write(metrics []telegraf.Metric) error {
+	ctx := context.Background()
+
+	for _, m := range metrics {
+		timeSeries := []*monitoringpb.TimeSeries{}
+
+		for k, v := range m.Fields() {
+			value, err := getStackdriverTypedValue(v)
+			if err != nil {
+				log.Printf("E! [output.stackdriver] get type failed: %s", err)
+				continue
+			}
+
+			metricKind, err := getStackdriverMetricKind(m.Type())
+			if err != nil {
+				log.Printf("E! [output.stackdriver] get metric failed: %s", err)
+				continue
+			}
+
+			timeInterval, err := getStackdriverTimeInterval(metricKind, StartTime, m.Time().Unix())
+			if err != nil {
+				log.Printf("E! [output.stackdriver] get time interval failed: %s", err)
+				continue
+			}
+
+			// Prepare an individual data point.
+			dataPoint := &monitoringpb.Point{
+				Interval: timeInterval,
+				Value:    value,
+			}
+
+			// Prepare time series.
+			timeSeries = append(timeSeries,
+				&monitoringpb.TimeSeries{
+					Metric: &metricpb.Metric{
+						Type:   path.Join("custom.googleapis.com", s.Namespace, m.Name(), k),
+						Labels: getStackdriverLabels(m.Tags()),
+					},
+					MetricKind: metricKind,
+					Resource: &monitoredrespb.MonitoredResource{
+						Type: "global",
+						Labels: map[string]string{
+							"project_id": s.Project,
+						},
+					},
+					Points: []*monitoringpb.Point{
+						dataPoint,
+					},
+				})
+		}
+
+		if len(timeSeries) < 1 {
+			continue
+		}
+
+		// Prepare time series request.
+		timeSeriesRequest := &monitoringpb.CreateTimeSeriesRequest{
+			Name:       monitoring.MetricProjectPath(s.Project),
+			TimeSeries: timeSeries,
+		}
+
+		// Create the time series in Stackdriver.
+		err := s.client.CreateTimeSeries(ctx, timeSeriesRequest)
+		if err != nil {
+			log.Printf("E! [output.stackdriver] create time series failed: %s", err)
+			continue
+		}
+	}
+
+	return nil
+}
+
+func getStackdriverTimeInterval(
+	m metricpb.MetricDescriptor_MetricKind,
+	start int64,
+	end int64,
+) (*monitoringpb.TimeInterval, error) {
+	switch m {
+	case metricpb.MetricDescriptor_GAUGE:
+		return &monitoringpb.TimeInterval{
+			EndTime: &googlepb.Timestamp{
+				Seconds: end,
+			},
+		}, nil
+	case metricpb.MetricDescriptor_CUMULATIVE:
+		return &monitoringpb.TimeInterval{
+			StartTime: &googlepb.Timestamp{
+				Seconds: start,
+			},
+			EndTime: &googlepb.Timestamp{
+				Seconds: end,
+			},
+		}, nil
+	case metricpb.MetricDescriptor_DELTA, metricpb.MetricDescriptor_METRIC_KIND_UNSPECIFIED:
+		fallthrough
+	default:
+		return nil, fmt.Errorf("unsupported metric kind %T", m)
+	}
+}
+
+func getStackdriverMetricKind(vt telegraf.ValueType) (metricpb.MetricDescriptor_MetricKind, error) {
+	switch vt {
+	case telegraf.Untyped:
+		return metricpb.MetricDescriptor_GAUGE, nil
+	case telegraf.Gauge:
+		return metricpb.MetricDescriptor_GAUGE, nil
+	case telegraf.Counter:
+		return metricpb.MetricDescriptor_CUMULATIVE, nil
+	case telegraf.Histogram, telegraf.Summary:
+		fallthrough
```

Comment:
Can we send these as GAUGE?

Reply:
@jamesmaidment What do you think about this item?

Reply:
@danielnelson Stackdriver does support a Distribution type that could work for one or both of these https://cloud.google.com/monitoring/api/ref_v3/rest/v3/TimeSeries#distribution

I haven't had a lot of cycles to look into it and so far haven't had this use case. I'm not sure if I can get to this immediately and could add it as a patch later?

Reply:
Yeah we can add proper support later.  Do you think we should send them as GAUGE values for now or do you think it is better to omit them as it is now?

Reply:
I think perhaps we should omit them for now and add proper support for Histogram & Summary using the Distribution type in a future release just to not cause having to change the behavior in future if people come to rely on the Gauge metric types.